### PR TITLE
BUG: MaskedArray ignores __array_priority__

### LINKED ISSUE #5227: MaskedArray ignores __array_priority__
Given a class like this:

```
class X(int):
    def __add__(self, other):
        return 42
    def __radd__(self, other):
        return 24
    __array_priority__ = 20
```

Adding to a numpy array works as expected:

```
>>> np.array(2) + X()
24
```

However, when I add `X()` to a masked array, `X.__radd__` is not called:

```
>>> np.ma.array(2) + X()
2
```

### PR DESCRIPTION
**Pull request by abalkin**
fixes: #5227 

This patch partially fixes the issue (I only covered addition, subtraction and multiplication).  If you agree that this is the right approach - I will add remaining operations.

### CODE REVIEW

File: numpy/ma/tests/test_core.py [hunk 1]
```diff
@@ -30,6 +30,11 @@
 class TestMaskedArray(TestCase):
     # Base test class for MaskedArrays.
 
+    # Python 2.6 compatibility
+    if not hasattr(TestCase, 'assertIsInstance'):
+        def assertIsInstance(self, obj, cls):
+            self.assertTrue(isinstance(obj, cls)
```
Comment (charris) [inline numpy/ma/tests/test_core.py]: Causes universal test failures.

**Comment by charris** [comment]
@abalkin  Could you finish this up?

File: numpy/ma/core.py [hunk 2]
```diff
@@ -901,22 +901,32 @@ class _MaskedBinaryOperation:
...
         self.f = mbfunc
         self.fillx = fillx
         self.filly = filly
+        self.op = op
         self.__doc__ = getattr(mbfunc, "__doc__", str(mbfunc))
         self.__name__ = getattr(mbfunc, "__name__", str(mbfunc))
         ufunc_domain[mbfunc] = None
         ufunc_fills[mbfunc] = (fillx, filly)
 
+    def __get__(self, obj, type=None):
```
Comment (mhvk) [inline numpy/ma/core.py]: I'm slightly confused why this is being done. It makes this a descriptor, right?

File: numpy/ma/core.py [hunk 3]
```diff
@@ -901,22 +901,32 @@ class _MaskedBinaryOperation:
...
+        return partial(self, obj)
+
     def __call__ (self, a, b, *args, **kwargs):
         "Execute the call behavior."
+        # Check array priority
+        if self.op:
```
Comment (mhvk) [inline numpy/ma/core.py]: In `ndarray` C code, `NotImplemented` is only returned if the array priority is higher *and* the other class defines the reverse operation. I think one could use a similar structure as you have here, except use `rop` to pass on the reverse operator and here do:
```
if self.rop and getattr(b, self.rop, None) and getattr(b, '__array_priority__', -1) > a.__array_priority__:
    return NotImplemented
```

File: numpy/ma/core.py (hunk 2 above)
Comment (mhvk) [inline numpy/ma/core.py]: Or is this what allows you to directly define `__add__` below?

**Comment by mhvk** [comment]
@abalkin - pointed to this by @charris -- this seems a good approach, and is probably the safest one can be.

In a slightly larger picture, though, I wonder if one isn't duplicating more of the `ufunc` machinery than necessary. I somewhat wonder why all the methods are overridden at all; with `__mul__` removed, the normal code would return `NotImplemented` and pass on to `X.__rmul__`. One can see this by using a function that is not defined, such as `__or__` (lucky some where missed!):
```
class X(int):
    __array_priority__ = 1000
    def __ror__(self, other):
        return 42

np.ma.array(1) | X(2)
# yields 42
```
Anyway, this mostly as a curiosity; I fear removing the methods would imply a great deal more work, and possibly a lot of rewriting of `__array_wrap__` -- which would suggest one might as well go for `__numpy_ufunc__` (which should make `MaskedArray` *much* simpler).

**Comment by abalkin** [comment]
@mhvk - I agree that `__numpy_ufunc__` is the solution, but I am not sure it is ready for prime time at this point.  On the other hand rewriting ma to use `__numpy_ufunc__` may serve as a good test-bed.

**Comment by mhvk** [comment]
@abalkin - from my work on `Quantity`, I think `__numpy_ufunc__` as currently implemented should do the trick; I don't think anything that is decided in #5844 would change that. That said, it would be quite a lot of work, and one of those cases where it is unclear whether one should start changing the code or just start from scratch and just ensure the new code passes all tests (which I think are decently complete for `MaskedArray`). I know I won't have time to do this myself anytime soon, but I'd be willing to help, both in thinking of approaches and in testing that it resolves the issues I have raised with the current implementation (#3907, #4586, #4617)

If you want to have some sense of what a `__numpy_ufunc__` method would look like, see https://github.com/astropy/astropy/pull/2948 and in particular the code starting at https://github.com/mhvk/astropy/blob/quantity-numpy-ufunc-additions/astropy/units/quantity.py#L409

**Comment by charris** [comment]
Is the suggestion that I can close this in favor of a `__numpy_ufunc__` approach at some future time? I confess that adding an `op` argument to standard python functons strikes me a bit odd.

**Comment by njsmith** [comment]
I believe that #5864 also addresses the issues here in a different way.

**Comment by mhvk** [comment]
Specifically, https://github.com/njsmith/numpy/commit/c3c819ae27c103b22ec0e7ce5924986a18b29c30
@abalkin - does that commit indeed resolve your problems as well?

**Comment by charris** [comment]
I'm going to close this, if someone wants to pursue this, please squeak.

@njsmith A short explanation of how #5864 would solve this would be useful.

**Comment by njsmith** [comment]
Currently there are two different mechanisms used to handle the interaction
of __array_priority__ and __binop__ dispatch: ndarray uses one (in
number.c), and np.ma uses the other (in umath). They work differently in
subtly ways, and this bug is about one of the places they disagree. The
main goal of #5864 is to clean up the ufunc code, so I had to switch np.ma
to stop using the ufunc stuff. Since I had to redo it anyway, I switched
np.ma to use the same __array_priority__ rule that ndarray already uses,
which has the side-effect of fixing this issue.

**Pull request closed without merging**